tools/ci/actions/check-configure.sh

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+#**************************************************************************
+#*                                                                        *
+#*                                 OCaml                                  *
+#*                                                                        *
+#*                 David Allsopp, OCaml Labs, Cambridge.                  *
+#*                                                                        *
+#*   Copyright 2021 David Allsopp Ltd.                                    *
+#*                                                                        *
+#*   All rights reserved.  This file is distributed under the terms of    *
+#*   the GNU Lesser General Public License version 2.1, with the          *
+#*   special exception on linking described in the file LICENSE.          *
+#*                                                                        *
+#**************************************************************************
+
+# Hygiene Checks: ensure that configure.ac generates configure
+# This tests both branches and PRs. Any commit which updates either files which
+# affect configure (configure.ac, VERSION, aclocal.m4 and build-aux/*) and also
+# which alter this script.
+# The behaviour is slightly different for pushes vs pull requests: in a PR, all
+# commits must be correct; in a push, it must be the case that the configure is
+# correct at the tip of the branch. This allows you to push a correcting PR to
+# trunk, for example, but rejects a PR which includes bad commits (for increased
+# bisect safety).
+
+set -e
+
+if [[ $1 = 'pull_request' ]]; then
+  ALL_COMMITS_MUST_PASS=1
+else
+  ALL_COMMITS_MUST_PASS=0
+fi
+
+# We need all the commits in the PR to be available
+. tools/ci/actions/deepen-fetch.sh
+
+# Display failing commits in red for PRs and yellow for branches (error/warning)
+if ((ALL_COMMITS_MUST_PASS)); then
+  COLOR='31'
+else
+  COLOR='33'
+fi
+
+CI_SCRIPT='tools/ci/actions/check-configure.sh'
+PATHS='configure\|configure\.ac\|VERSION\|aclocal\.m4\|build-aux/.*'
+
+# $1 - commit to checkout files from
+# $2 - range of commits to diff
+# When testing a single commit, $1 and $2 will be the same; when validating the
+# tip of a branch, $1 will be HEAD and $2 will be the range of commits in the
+# branch.
```

Comment:
The way important git data is passed as parameters to the script, and then referred to by numeric positions makes them fairly hard to read and, I would guess, painful to modify. (If you keep this style, I think you should just name all parameters $1..$n to a readable name at the entry of the script, and never use by-position parameters in the rest of the script.)

```

          tools/ci/actions/check-configure.sh
          '${{ github.event_name }}'
          '${{ github.event.pull_request.base.ref }}'
          '${{ github.event.pull_request.base.sha }}'
          '${{ github.event.pull_request.head.ref }}'
          '${{ github.event.pull_request.head.sha }}'
          '${{ github.event.ref }}'
          '${{ github.event.before }}'
          '${{ github.event.ref }}'
          '${{ github.event.after }}'
```

```
[...]
    git checkout -q "$1"
[...]
```

I have two questions here:
- Why is the case that some variables keep a civil name? I see `$MERGE_BASE` and `$PR_HEAD` here and `$commit` and `$COMMIT_RANGE`. Ah... (answering my own question) they are defined by `deepen-fetch`, which does come with a very nice comment explaining the interface it exports.
- Is there a better way? In the `run` actions, could you set environment variables with those many git-relevant values, and then use the variable names instead of `$3` in the scripts?

Reply:
Hah - the ones which remain have come from `travis-ci.sh`... those aren't script parameters, they're parameters in the functions, which rather proves your point!

Reply:
I've intentionally not used environment variables for this - it's routinely a nightmare with the Windows runners (on AppVeyor) and using environment variables to control things like this I find even harder to reason about than the scripts (because logs do not make clear the value of the environment on every call).

However, GitHub Actions also doesn't support a ternary operator (as noted in `deepen-fetch.sh` L27) which means that even if it is possible to express that environment block correctly, it will be YAML hell to figure out how to write the matrix for it!

Reply:
I think I understand what's going on now (I forced myself to read the code).

We need 4 variables whose value should depend on the request type (whether we are in a pull request or in a direct push). The previous code had something like `if request==pull then FOO=X else FOO=Y`. The way you encode this is with 9 variables, $1 to tell the request type, and then the variables $2..$5 in the first case and $6..$9 in the second case. (But one test only gets 5 variables because we statically know the value of $1 so we know $6..$9 are not needed.)

`deepen-fetch` is in charge of deciding which window of 4 variables to use, and it re-exports them under readable names for the rest of the script to use. (But it also drops $1, with the consequence that some code needs to run before `deepen-fetch` to access it for the rest of the script.) The rest of the script never access $2-$9, only the named variables exported by deepen-fetch.

(I think it would have made sense to separate two aspects of deepen-fetch, one script that properly defines the value of variables depending on the request type (ideally, also re-exporting $1 to allow this to be done immediately on script entry), and another script that runs the exponential-growth merge-base fetching loop that we need because the world is a dark place. But oh well, it's fun to read about exponential merge-base fetching to understand how variables are defined.)

Then the $1..$2 I was pointing at are not actually arguments of the script itself, but arguments of local CheckFoo functions. They follow a common protocol that $1 is a name of commit to check, and $2 is the range of commits whose changed files should be checked (but at the state they are in $1). The Configure check was extracted from the Typo check to make things easier, but actually it never uses $2. Finally, after my last comment (I think?) you gave readable names to $1 and $2 on each function entry, so now the use of variables is very readable -- if somewhat complex.